keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -0,0 +1,464 @@
+---
+title: Immutable Fields
+authors:
+  - "@apelisse"
+  - "@sttts"
+owning-sig: sig-api-machinery
+participating-sigs:
+  - sig-api-machinery
+reviewers:
+  - "@erictune"
+  - "@jpbetz"
+  - "@liggitt"
+  - "@logicalhan"
+  - "@p0lyn0mial"
+approvers:
+  - "@liggitt"
+  - "@deads2k"
+editor: "@sttts"
+creation-date: 2019-06-03
+last-updated: 2019-10-01
+status: provisional
+see-also:
+  - "/keps/sig-api-machinery/0006-apply.md"
+---
+
+# Immutable fields
+
+## Release Signoff Checklist
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+A lot of fields in APIs tend to be "immutable", they can't be changed after
+creation. This is true for example for many of the fields in pods. There is
+currently no way to declaratively specify that fields are immutable, and one
+has to rely on either built-in validation for core types, or have to build a
+validating webhooks for CRDs.
+
+Providing a new `// +immutable` marker would help 
+- to make the API more descriptive to users
+- to help API developers by verifying these assertions automatically
+- and to publish this information via OpenAPI.
+
+## Motivation
+
+There are resources in Kubernetes which have immutable fields by design,
+i.e. after creation of an object, those fields cannot be mutated anymore. E.g. a
+pod's specification is mostly unchangeable once it is created. To change the
+pod, it must be deleted, recreated and rescheduled. Users want to implement the
+same kind of read-only semantics for CustomResources, for example:
+https://github.com/kubernetes/kubernetes/issues/65973. Today this is only possible
+with the unreasonable development overhead of a webhook.
```

Comment:
I have a vague memory of an immutable-once-set semantic we wanted at one point (would enable an async allocator to set a field once via the API, after which it would be locked). Was that discarded or deferred? Not having that means any immutable field must be set synchronously on create by the object originator or a webhook.

Reply:
Am not aware of that old discussion, but it definitly makes sense to flesh that out. 

It sounds like you want a way to declare a struct not completely immutable, but stop it from being deleted, and certain fields to be set-once-only:
- This is more than marking fields immutable, because you could still set them to undefined. 
- It is less than marking the whole struct's keys immutable because not all fields might be set only.
In other words, this is some weaker form of `x-kubernetes-keys` on the hosting struct, let's call it `x-kubernetes-keys-set-once`:

Example:

```yaml
apiVersion: setonce/v1
kind: Once
spec:
   once: true
   name: foo
```

with

```yaml
type: object
x-kubernetes-immutable-keys: true
properties:
  spec:
    type: object
    x-kubernetes-keys-set-once: ["once","name"]
    properties:
      once:
        type: boolean
      name:
        type: string
        x-kubernetes-immutable: true
```

Here, `name` can only be set to a fixed value once, and never changed anymore. `once` can be set once (i.e. from undefined to some value), but the actual value can be changed any time.

Reply:
@liggitt Did you have an actual use-case in mind? I can't remember talking about that either, and I'd really like to keep the number of knobs low if we can.

Reply:
> Did you have an actual use-case in mind?

We use this capability in built-in resources today:
* scheduler setting spec.nodeName (and it being immutable after that point)
* populating clusterIP in services
* append-only lists (like ephemeralContainers)
* remove-only lists (like finalizers that may only shrink after deletion)

Talked through a lot of scenarios with @sttts, and I think the API guideline to [think twice about boolean use](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#primitive-types) probably applies here. Modeling restrictions on key/value immutability as an enum rather than a boolean allows more use cases to avoid requiring a validating webhook. Something like:

```
x-kubernetes-mutability: Immutable | AddOnly | RemoveOnly
x-kubernetes-key-mutability: Immutable | AddOnly | RemoveOnly
```

Reply:
> scheduler setting spec.nodeName (and it being immutable after that point)

This is write-once, I agree (but not sure we need a general mechanism for this).

> populating clusterIP in services

Is this actually write-once?

> append-only lists (like ephemeralContainers)

append-only isn't quite the same as write-once.

> remove-only lists (like finalizers that may only shrink after deletion)

IMO the "after deletion" qualifier definitely means that there's not going to be a general mechanism for this.

Reply:
>> scheduler setting spec.nodeName (and it being immutable after that point)
>
> This is write-once, I agree (but not sure we need a general mechanism for this).
>
>> populating clusterIP in services
>
> Is this actually write-once?

We set clusterIPs in-process in the service registry implementation, and do not allow it to change on update. To replicate these interactions without write-once capabilities, a custom resource would either have to:
* use a mutating webhook on create (which would require allocation to be really fast) and built-in immutability on the CR schema
* use a controller to allocate asynchronously and a validating webhook to prevent changing/unsetting a value

If we can allow custom resources to express invariants without hooking up dynamic logic, that seems valuable.

> append-only isn't quite the same as write-once

You can accomplish append-only with a combination of `key-mutability: AddOnly` on the container, and `mutability: Immutable` on the individual items.

> IMO the "after deletion" qualifier definitely means that there's not going to be a general mechanism for this.

I agree the AddOnly case is stronger than RemoveOnly.

Reply:
> If we can allow custom resources to express invariants without hooking up dynamic logic, that seems valuable.

Yes, the question isn't *if* it is valuable, the question is how valuable, to how many users, and at what cost, where cost is initial implementation cost (amortized) + ongoing maintenance cost + ongoing API complexity cost to users and us